src/machines.jl

```diff
@@ -78,39 +79,204 @@ constructor for retraining on new data using the saved model.
 """
 function save(file::Union{String,IO},
               mach::Machine;
-              verbosity=1,
-              format=:julia_serialize,
-              compression=:none,
               kwargs...)
     isdefined(mach, :fitresult)  ||
         error("Cannot save an untrained machine. ")
 
-    # fallback `save` method returns `mach.fitresult` and saves nothing:
-    serializable_fitresult =
-        save(_filename(file), mach.model, mach.fitresult; kwargs...)
+    smach = serializable(file, mach, kwargs...)
+
+    serialize(file, smach)
+end
+
+"""
+Not sure how to provide new arguments:
+    - change mach.data?
+"""
+function machine(file::Union{String,IO}, args...)
+    smach = deserialize(file)
+    restore!(smach, file)
+end
+
+
+###############################################################################
+#####                           UTILITIES                                 #####
+###############################################################################
+
+newcache(cache::NamedTuple) = Base.structdiff(cache, NamedTuple{(:data,)})
+newcache(cache) = cache
+
+wipe_cached_data!(mach1::Machine, mach2::Machine) = 
+    mach1.cache = newcache(mach2.cache)
+
+setreport!(mach::Machine, report) = 
+    setfield!(mach, :report, report)
+
+function setreport!(mach::Machine{<:Composite}, report)
+    glb_node = glb(mach)
+    mach.report = merge(MLJBase.report(glb_node), MLJBase.report_additions(mach.fitresult))
+end
+
+function add_exception_nodes!(d::IdDict, nodes::Vararg{AbstractNode}) 
+    for node in nodes
+        if node isa MLJBase.ErrorNode
+            d[node] = node
+        end
+    end
+end
+
+###############################################################################
+#####         PROBABLY TO BE EXPORTED TO THEIR RESP MODULES               #####
+###############################################################################
+
+using MLJTuning
+using MLJEnsembles
+
+
+MLJModelInterface.save(filename, model::MLJTuning.EitherTunedModel, fitresult::Machine; kwargs...) =
+    serializable(filename, fitresult, kwargs...)
+
+function MLJModelInterface.restore(filename, model::MLJTuning.EitherTunedModel, fitresult)
+    fitresult.fitresult = restore(filename, fitresult.model, fitresult.fitresult)
+    return fitresult
+end
+
+MLJModelInterface.save(filename, model::MLJEnsembles.EitherEnsembleModel, fitresult; kwargs...) =
+    MLJEnsembles.WrappedEnsemble(
+        fitresult.atom,
+        [save(filename, fitresult.atom, fr, kwargs...) for fr in fitresult.ensemble]
+    )
+
+function MLJModelInterface.restore(filename, model::MLJEnsembles.EitherEnsembleModel, fitresult)
+    return MLJEnsembles.WrappedEnsemble(
+        fitresult.atom,
+        [restore(filename, fitresult.atom, fr) for fr in fitresult.ensemble]
+    )
+end
+
+
+"""
+    save(filename, model::Composite, fitresult; kwargs...)
+"""
+function save(filename, model::Composite, fitresult; kwargs...)
+    # THIS IS WIP: NOT WORKING
```

Comment:
I wonder if rather than duplicating code in MLJBase we can make a (non-breaking) enhancement to the function defined there (with a new keyword) so it does what you need??